src/space.c

```diff
@@ -2792,11 +2872,127 @@ static int system_parseAsteroidField( const xmlNodePtr node, StarSystem *sys )
-   a->nb = floor( ABS(a->aera) / 100000 * a->density );
+   a->nb      = floor( ABS(a->aera) / 500000 * a->density );
+   a->ndebris = floor(100*a->density);
 
    /* Added asteroid. */
    sys->nasteroids++;
 
+   /* Initialize the convex subsets. */
+   a->subsets = malloc( sizeof(AsteroidSubset) );
+   a->subsets[0].ncorners = a->ncorners;
+   a->subsets[0].corners = a->corners;
+   a->nsubsets = 1;
+
+   /* Cut the subsets until they are all convex. */
+   retry  = 1;
+   getout = 0;
+
+   while (retry) {
+      retry = 0;
+      for (i=0; i<a->nsubsets; i++) {
+         sub = &a->subsets[i];
+         for (j=0; j<sub->ncorners; j++) {
+            /* Find the 2 other corners. */
+            if (j>0 && j<sub->ncorners-1) {
+               k = j-1;
+               l = j+1;
+            }
+            else if (j==0) {
+               k = sub->ncorners-1;
+               l = j+1;
+            }
+            else { /* (j==ncorners-1) */
+               k = j-1;
+               l = 0;
+            }
+
+            /* Test the orientation of the angle towards signed aera. */
+            pro = (sub->corners[k].x-sub->corners[j].x) *
+                  (sub->corners[l].y-sub->corners[j].y) - 
+                  (sub->corners[k].y-sub->corners[j].y) *
+                  (sub->corners[l].x-sub->corners[j].x);
+
+            if (pro * a->aera > 0) {
+               /* Cut here, find an other point. */
+               x = (sub->corners[k].x + sub->corners[l].x)/2;
+               y = (sub->corners[k].y + sub->corners[l].y)/2;
+
+               pro = 1.;
+               for (m=0; m<sub->ncorners; m++) {
+                  prob = (x-sub->corners[j].x)*(sub->corners[m].x-sub->corners[j].x) +
+                         (y-sub->corners[j].y)*(sub->corners[m].y-sub->corners[j].y);
+                  if ( prob < pro ) {
+                     pro = prob;
+                     n = m;
+                  }
+               }
+
+               if (pro>=0)
+                  WARN("Non-convex asteroid polygon seems to be self-intersecting in %d", sys->name);
+
+               /* Split subset i into 2 subsets */
+
+               a->subsets = realloc( a->subsets, (a->nsubsets+1)*sizeof(AsteroidSubset) );
+
+               /* Need to recall who is sub. */
+               sub = &a->subsets[i];
+               newsub = &a->subsets[a->nsubsets];
+
+               /* First, make sure n<j: invert j and n if needed */
+               if (j<n) {k=n; n=j; j=k;}
+
+               newsub->ncorners = j-n+1;
+               newsub->corners  = malloc( sizeof(Vector2d)*newsub->ncorners );
+
+               /* Add some nodes to the new subset */
+               k = 0;
+               for (m=n; m<j+1; m++) {
+                  newsub->corners[k] = sub->corners[m];
+                  k++;
+               }
+
+               /* Remove some nodes to the old subset */
+               memmove(&sub->corners[n+1], &sub->corners[j], sizeof(AsteroidSubset)*(j-n-1));
```

Comment:
Trying the code, it seems like there is an invalid read here. 8 bytes passed the allocated area, so maybe an off-by-one miss?